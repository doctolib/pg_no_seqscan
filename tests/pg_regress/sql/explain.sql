-- Test that EXPLAIN (COSTS OFF) queries are ignored
LOAD 'pg_no_seqscan';
SET pg_no_seqscan.level = ERROR;
CREATE TABLE test_explain AS (SELECT * FROM generate_series(1,10) AS id);

-- EXPLAIN (COSTS OFF) should not trigger errors
EXPLAIN (COSTS OFF) SELECT * FROM test_explain;

-- EXPLAIN (COSTS OFF) ANALYZE should not trigger errors
EXPLAIN (ANALYZE, COSTS OFF, TIMING OFF, SUMMARY OFF) SELECT * FROM test_explain;

-- But regular query should trigger error
SELECT * FROM test_explain;

-- cleanup
drop table test_explain;