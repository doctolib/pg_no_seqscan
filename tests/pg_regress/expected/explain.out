-- Test that EXPLAIN (COSTS OFF) queries are ignored
LOAD 'pg_no_seqscan';
SET pg_no_seqscan.level = ERROR;
-- To produce stable regression, inspired from:
-- https://github.com/postgres/postgres/blob/master/src/test/regress/sql/explain.sql
create function explain_filter(text) returns setof text
language plpgsql as
$$
declare
ln text;
begin
for ln in execute $1
    loop
        -- Replace any numeric word with just 'N'
        ln := regexp_replace(ln, '-?\m\d+\M', 'N', 'g');
        -- In sort output, the above won't match units-suffixed numbers
        ln := regexp_replace(ln, '\m\d+kB', 'NkB', 'g');
        -- Ignore text-mode buffers output because it varies depending
        -- on the system state
CONTINUE WHEN (ln ~ ' +Buffers: .*');
        -- Ignore text-mode "Planning:" line because whether it's output
        -- varies depending on the system state
CONTINUE WHEN (ln = 'Planning:');
        return next ln;
end loop;
end;
$$;
CREATE TABLE test_explain AS (SELECT * FROM generate_series(1,10) AS id);
-- EXPLAIN (COSTS OFF) should not trigger errors
EXPLAIN (COSTS OFF) SELECT * FROM test_explain;
        QUERY PLAN        
--------------------------
 Seq Scan on test_explain
(1 row)

-- EXPLAIN (COSTS OFF) ANALYZE should not trigger errors
select explain_filter('EXPLAIN (ANALYZE, COSTS OFF, TIMING OFF, SUMMARY OFF, BUFFERS OFF) SELECT * FROM test_explain;');
                  explain_filter                  
--------------------------------------------------
 Seq Scan on test_explain (actual rows=N loops=N)
(1 row)

-- But regular query should trigger error
SELECT * FROM test_explain;
ERROR:  A 'Sequential Scan' on test_explain has been detected.
  - Run an EXPLAIN on your query to check the query plan.
  - Make sure the query is compatible with the existing indexes.

Query: SELECT * FROM test_explain;

-- cleanup
drop table test_explain;
